index.js

```diff
@@ -254,3 +255,152 @@ module.exports.node = (scriptPath, args, options = {}) => {
+	} catch (error) {
+		const errorPromise = Promise.reject(makeError({
+			error,
+			stdout: undefined,
+			stderr: undefined,
+			all: undefined,
+			command,
+			parsed,
+			timedOut: false,
+			isCanceled: false,
+			killed: false
+		}));
+
+		// We don't require the caller to handle the rejection since errors
+		// will be translated into pipe/pipeline failure
+		errorPromise.catch(() => {});
+
+		const dummyDuplex = new stream.Duplex({
+			read() {
+				this.destroy(error);
+			},
+			write(chunk, encoding, callback) {
+				callback(error);
+			}
+		});
+
+		return mergePromise(dummyDuplex, errorPromise);
+	}
+
+	const spawnedPromise = getSpawnedPromise(spawned);
+	const timedPromise = setupTimeout(spawned, parsed.options, spawnedPromise);
+	const processDone = setExitHandler(spawned, parsed.options, timedPromise)
+		.catch(error => ({error, signal: error.signal, timedOut: error.timedOut}));
+
+	const context = {isCanceled: false};
+
+	spawned.kill = spawnedKill.bind(null, spawned.kill.bind(spawned));
+	spawned.cancel = spawnedCancel.bind(null, spawned, context);
+
+	crossSpawn._enoent.hookChildProcess(spawned, parsed.parsed);
+
+	const all = makeAllStream(spawned, parsed.options);
+	const {stdin} = spawned;
+	const output = all ? all : spawned.stdout;
+
+	const duplex = new stream.Duplex({
+		autoDestroy: false,
+		write(chunk, encoding, callback) {
+			stdin.write(chunk, encoding, callback);
+		},
+		final(callback) {
+			stdin.end(callback);
+		},
+		read() {
+			output.resume();
+		},
+		destroy(error, callback) {
+			if (spawned.exitCode === null && spawned.signalCode === null) {
+				spawned.once('exit', () => {
+					callback(error);
+				});
+
+				spawned.once('error', () => {
+					callback(error);
+				});
+
+				spawned.cancel();
+			} else {
+				callback(error);
+			}
+		}
+	});
+
+	stdin.once('error', error => {
+		duplex.destroy(error);
+	});
+
+	output.once('error', error => {
+		duplex.destroy(error);
+	});
+
+	output.once('end', async () => {
+		const {error, exitCode, signal} = await processDone;
+		if (!error && exitCode === 0 && signal === null) {
+			duplex.push(null);
+		}
+	});
+	output.on('data', chunk => {
+		if (!duplex.push(chunk)) {
+			output.pause();
+		}
+	});
+
+	const result = processDone.then(({error, exitCode, signal, timedOut}) => {
+		if (error || exitCode !== 0 || signal !== null) {
+			const returnedError = makeError({
+				error,
+				exitCode,
+				signal,
+				stdout: undefined,
+				stderr: undefined,
+				all: undefined,
+				command,
+				parsed,
+				timedOut,
+				isCanceled: context.isCanceled,
+				killed: spawned.killed
+			});
+
+			throw returnedError;
+		}
+
+		return {
+			command,
+			exitCode: 0,
+			stdout: undefined,
+			stderr: undefined,
+			all: undefined,
+			failed: false,
+			timedOut: false,
+			isCanceled: false,
+			killed: false
+		};
+	});
+
+	Promise.all([
+		result.then(() => {}).catch(error => error),
```

Comment:
Do you really need `.then` here?

Reply:
Yes, otherwise non-error result will be resolved by the complete promise. A more explicit intent would be

`result.then(() => { return null; }).catch(error => error)`

But anyway this is changed by the use of await instead of then.